### PR TITLE
borgmatic: 1.7.15 -> 1.8.1

### PR DESCRIPTION
**Pull request by imlonghao**
## Description of changes

1. Bump version from 1.7.15 to 1.8.1, change log can be found at

- https://projects.torsion.org/borgmatic-collective/borgmatic/releases/tag/1.8.0
- https://projects.torsion.org/borgmatic-collective/borgmatic/releases/tag/1.8.1

2. move `source_directories` and `repositories` out of `location` scope, or else there is a warning from borgmatic, see https://projects.torsion.org/borgmatic-collective/borgmatic/issues/721

```
/etc/borgmatic.d/a.yaml: Configuration sections (like location: and storage:) are deprecated and support will be removed from a future release. To prepare for this, move your options out of sections to the global scope.
```

3. make `repositories` to a list of key/value pairs, or else there is a warning from borgmatic, see https://projects.torsion.org/borgmatic-collective/borgmatic/issues/635

```
/etc/borgmatic.d/a.yaml: The repositories option now expects a list of key/value pairs. Lists of strings for this option are deprecated and support will be removed from a future release.
```

4. add a warning if users are still using `location` section, old config for example

```
services.borgmatic = {
  enable = true;
  configurations = {
    filebrowser = {
      location = {
        source_directories = [
          "/path/to/filebrowser"
        ];
        repositories = [
          "ssh://x@x.repo.borgbase.com/./repo"
        ];
      };
    };
  };
}
```

should be upgraded to

```
services.borgmatic = {
  enable = true;
  configurations = {
    filebrowser = {
      source_directories = [
        "/persist/docker/filebrowser"
      ];
      repositories = [
        { path="ssh://x@x.repo.borgbase.com/./repo"; label="borgbase"; }
      ];
    };
  };
}
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by raboof** [comment]
@imlonghao would you like to fix the conflict?

File: nixos/modules/services/backup/borgmatic.nix [hunk 1]
```diff
@@ -62,6 +80,13 @@ in
 
   config = mkIf cfg.enable {
 
+    warnings = []
+      ++ optional (cfg.settings != null && cfg.settings.location != null)
+        "services.borgmatic.settings.`location` is deprecated, please move your options out of sections to the global scope"
```
Comment (raboof) [inline nixos/modules/services/backup/borgmatic.nix:85]: why is only `location` in backticks?

**Review by raboof — APPROVED** [review]
Looks good, one question around the formatting of the warnings

File: nixos/modules/services/backup/borgmatic.nix (hunk 1 above)
Comment (imlonghao) [reply nixos/modules/services/backup/borgmatic.nix]: Should be a typo, updated.

**Pull request merged**